crates/uv-extract/src/stream.rs

```diff
@@ -143,6 +144,16 @@ async fn untar_in(
     mut archive: tokio_tar::Archive<&'_ mut (dyn tokio::io::AsyncRead + Unpin)>,
     dst: &Path,
 ) -> std::io::Result<()> {
+    // Like `tokio-tar`, canonicalize the destination prior to unpacking.
+    let dst = fs_err::tokio::canonicalize(dst).await?;
+
+    // Memoize filesystem calls to canonicalize paths.
+    let mut memo = FxHashSet::default();
+
+    // Delay any directory entries until the end, to ensure that directory permissions do not
+    // interfere with descendant extraction.
```

Comment:
Is this a concern for us? We usually create directories from archives without preserving permissions (at least in the zip path we should).

Reply:
Also maybe this is me being unfamiliar with the codebase but this is so... bizarre. How does this unpacking a file before the dir that contains it make sense? Is the idea that we create_dir_all any file path anyway, so the dir entries only create empty dirs and set perms on dirs that have files?

Reply:
For zips we always create the parent of a file, as a zip can contain directory entries, but sometimes there are no directory entries, just files.

https://github.com/astral-sh/uv/blob/4d3809cc6b28d71f26a782929b994abfebd7973c/crates/uv-extract/src/stream.rs#L66-L76

Reply:
> Is the idea that we create_dir_all any file path anyway, so the dir entries only create empty dirs and set perms on dirs that have files?

Yes. When we create a file, we create the directories required for it. This happens within the `async-tar` crate. The reason we have to do it "manually" here is because we have our own `unpack` that replicates the `async-tar` logic but applies different permissions.

Reply:
(I will look at why the `tar` crate introduced this before merging.)

Reply:
It kind of seems like a bug that we aren't preserving directory permissions in the zip case?

Reply:
personally, i'd prefer only preserving the executable bit and otherwise using the default umask; i'd find it odd if the permissions on the build host should determine the permissions in the deployment target.

Reply:
Yeah. Probably requires more invasive changes to the `tar` crate though.

Reply:
I'd like to review what pip does here before making other changes (since this is just preserving our status quo).